File: backend/app/models/booking.py
from sqlalchemy import ForeignKey, Integer, DateTime, Enum, String
from sqlalchemy.orm import Mapped, mapped_column, relationship
from datetime import datetime
import enum
import uuid

from core.database import Base


class BookingStatus(str, enum.Enum):
    ACTIVE = "active"
    CANCELLED = "cancelled"
    LATE_CANCEL = "late_cancel"
    STUDIO_CANCELLED = "studio_cancelled"


class Booking(Base):
    __tablename__ = "bookings"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    slot_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("slots.id"), nullable=False)
    client_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("clients.id"), nullable=False)
    seats_count: Mapped[int] = mapped_column(Integer, nullable=False)
    rental_count: Mapped[int] = mapped_column(Integer, default=0)
    status: Mapped[BookingStatus] = mapped_column(Enum(BookingStatus), default=BookingStatus.ACTIVE)
    price_total: Mapped[int] = mapped_column(Integer, nullable=False)
    allergy_comment: Mapped[str] = mapped_column(String(500), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    cancelled_at: Mapped[datetime] = mapped_column(DateTime, nullable=True)
    cancellation_reason: Mapped[str] = mapped_column(String(200), nullable=True)

    slot = relationship("Slot", back_populates="bookings")
    client = relationship("Client", back_populates="bookings")

File: android/app/src/main/java/com/mimokassy/data/models/Booking.kt
package com.mimokassy.data.models

import java.util.Date

enum class BookingStatus {
    ACTIVE, CANCELLED, LATE_CANCEL, STUDIO_CANCELLED
}

data class Booking(
    val id: String,
    val slotId: String,
    val clientId: String,
    val seatsCount: Int,
    val rentalCount: Int,
    val status: BookingStatus,
    val priceTotal: Int,
    val allergyComment: String?,
    val createdAt: String,
    val cancelledAt: String?,
    val cancellationReason: String?,
    val slot: Slot?
) {
    val isUpcoming: Boolean
        get() = status == BookingStatus.ACTIVE && slot?.startAt?.after(Date()) == true

    val isPast: Boolean
        get() = slot?.startAt?.before(Date()) == true

    val ownSeatsCount: Int
        get() = seatsCount - rentalCount

    val displayStatus: String
        get() = if (!status.isActive() && slot?.startAt?.before(Date()) == true) {
            "прошедшая"
        } else {
            status.displayName
        }

    val canCancel: Boolean
        get() = status == BookingStatus.ACTIVE && slot?.startAt?.after(Date()) == true
}

fun BookingStatus.isActive(): Boolean = this == BookingStatus.ACTIVE

val BookingStatus.displayName: String
    get() = when (this) {
        BookingStatus.ACTIVE -> "активна"
        BookingStatus.CANCELLED -> "отменена"
        BookingStatus.LATE_CANCEL -> "поздняя отмена"
        BookingStatus.STUDIO_CANCELLED -> "отменена студией"
    }
